File: backend/app/models/assigned_activity.py
"""AssignedActivity model.

A cognitive *activity* a care-team member (doctor/therapist) assigns to one of
their patients from a small set of safe, predefined templates. The patient sees
it in the mobile app and can start/complete it.

MEDICAL SAFETY: activities are cognitive **exercises only**. There is NO
diagnosis, treatment recommendation, prediction, or risk/mental-health scoring.
`generated_content` holds safe exercise parameters (e.g. word lists, sequence
length, number of rounds) produced from a fixed template — never free-form or
model-generated medical text.
"""

import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import (
    JSON,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    Uuid,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.db.mixins import UUIDPrimaryKeyMixin, utcnow

# Activity lifecycle statuses.
STATUS_ASSIGNED = "assigned"
STATUS_COMPLETED = "completed"
STATUS_SKIPPED = "skipped"


class AssignedActivity(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "assigned_activities"

    # The patient the activity is for (their profile).
    patient_profile_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("patient_profiles.id"), nullable=False
    )
    # The care-team member (doctor/therapist) who assigned it.
    assigned_by_user_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("users.id"), nullable=False
    )

    # One of the predefined template types (see modules/activities/templates.py).
    template_type: Mapped[str] = mapped_column(String(64), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    # Plain, supportive, non-diagnostic instructions.
    instructions: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    difficulty: Mapped[str] = mapped_column(String(32), nullable=False)
    duration_minutes: Mapped[int] = mapped_column(Integer, nullable=False)

    # assigned / completed / skipped.
    status: Mapped[str] = mapped_column(
        String(16), default=STATUS_ASSIGNED, nullable=False
    )
    # Safe, template-derived exercise parameters only (no medical content).
    generated_content: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utcnow, nullable=False
    )
    completed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    def __repr__(self) -> str:  # pragma: no cover - debug helper
        return (
            f"<AssignedActivity id={self.id!r} template={self.template_type!r} "
            f"patient={self.patient_profile_id!r} status={self.status!r}>"
        )

File: mobile/lib/features/activities/data/assigned_activity.dart
/// An activity the care team assigned to the patient
/// (from GET /api/v1/activities/my).
///
/// Cognitive-exercise content only — no diagnosis, treatment, prediction, or
/// scoring. `generatedContent` holds safe template parameters.
class AssignedActivity {
  const AssignedActivity({
    required this.id,
    required this.templateType,
    required this.title,
    required this.difficulty,
    required this.durationMinutes,
    required this.status,
    this.instructions,
    this.generatedContent,
    this.createdAt,
    this.completedAt,
  });

  final String id;
  final String templateType;
  final String title;
  final String? instructions;
  final String difficulty;
  final int durationMinutes;
  final String status; // assigned / completed / skipped
  final Map<String, dynamic>? generatedContent;
  final DateTime? createdAt;
  final DateTime? completedAt;

  bool get isCompleted => status == 'completed';
  bool get isSkipped => status == 'skipped';
  bool get isPending => status == 'assigned';

  /// The in-app game route for this template, or null if it is preview-only.
  String? get gameRoute {
    switch (templateType) {
      case 'memory_recall':
        return '/games/play/memory-recall';
      case 'attention_focus':
        return '/games/play/attention-focus';
      case 'reaction_time':
        return '/games/play/reaction-time';
      case 'sequence_recall':
        return '/games/play/sequence-order';
      case 'matching_game':
        return '/games/play/memory-match';
      default:
        return null; // daily_orientation and anything else: guided preview
    }
  }

  bool get isPlayable => gameRoute != null;

  factory AssignedActivity.fromJson(Map<String, dynamic> json) {
    DateTime? parse(Object? s) =>
        s is String ? DateTime.tryParse(s) : null;
    return AssignedActivity(
      id: (json['id'] ?? '').toString(),
      templateType: (json['template_type'] as String?) ?? '',
      title: (json['title'] as String?) ?? '',
      instructions: json['instructions'] as String?,
      difficulty: (json['difficulty'] as String?) ?? '',
      durationMinutes: (json['duration_minutes'] as num?)?.toInt() ?? 0,
      status: (json['status'] as String?) ?? 'assigned',
      generatedContent:
          (json['generated_content'] as Map?)?.cast<String, dynamic>(),
      createdAt: parse(json['created_at']),
      completedAt: parse(json['completed_at']),
    );
  }

  AssignedActivity copyWith({String? status, DateTime? completedAt}) {
    return AssignedActivity(
      id: id,
      templateType: templateType,
      title: title,
      instructions: instructions,
      difficulty: difficulty,
      durationMinutes: durationMinutes,
      status: status ?? this.status,
      generatedContent: generatedContent,
      createdAt: createdAt,
      completedAt: completedAt ?? this.completedAt,
    );
  }
}
